### PR TITLE
Add basis-template operators for mapmaking

### PR DESCRIPTION
**Pull request by sbiquard**
Replaces the per-family template operators with a single reusable "basis" abstraction, and wires it into the single-observation mapmaker. Two main commits: one adds the new operators + config + tests, the second migrates the mapmaker onto them and deletes the previous code.

## Motivation

The template machinery had one bespoke `*TemplateOperator` class per family (polynomial drift, scan-synchronous, HWP-synchronous, etc.), each re-implementing its own `mv`/transpose. Instead, this PR factors the shared structure into a small set of `Basis` types and one `PerDetectorTemplate` wrapper, so each family is a simple factory method.

Crucially, the new design will allow us to use `PerDetectorTemplate` easily for multi-observation mapmaking: its fixed-shape array leaves can be stacked along an observation axis, and make padding straightforward.

## What's new

`Basis`: a set of template functions that model a time-domain signal. `expand` synthesises the signal from amplitudes, `project` is its adjoint. Four different flavours trade memory for structure:
- `TensorBasis`: dense, stores every basis function value (with optional decimation to save memory).
- `KroneckerBasis`: factorised product of independent legs (e.g. azimuth × HWP); memory scales as the sum of leg sizes, not the product.
- `SegmentedBasis`: one segment per sample (e.g. per-scan-interval polynomials), stored sparsely.

`PerDetectorTemplate`: wraps a single-detector basis into a per-detector operator. Factory methods cover existing families (and more): `polynomial`, `scan_synchronous`, `binaz_synchronous`, `hwp_synchronous`, `azhwp_synchronous`, `binazhwp_synchronous`, `temperature` (T2P, with optional band-pass and decimation).

NB: T2P template operator not yet wired into any mapmaker (only makes sense with demodulated data).

Configuration: template families now configured via new `PolynomialOrders` / `BinsConfig` when applicable. Here is an example:

```yaml
# BEFORE (old keys)
templates:
  polynomial:
    max_poly_order: 3
  scan_synchronous:
    min_poly_order: 3
    max_poly_order: 7
  hwp_synchronous:
    n_harmonics: 3
  azhwp_synchronous:
    n_polynomials: 4   # orders 0..3
    n_harmonics: 4
    split_scans: false
  binazhwp_synchronous:
    n_azimuth_bins: 4
    n_harmonics: 4
    interpolate_azimuth: false
    smooth_interpolation: false

# AFTER (legendre / bins blocks)
templates:
  polynomial:
    legendre: { min_order: 0, max_order: 3 }
  scan_synchronous:
    legendre: { min_order: 3, max_order: 7 }
  hwp_synchronous:
    n_harmonics: 3
  azhwp_synchronous:
    legendre: { min_order: 0, max_order: 3 }   # was n_polynomials: 4
    n_harmonics: 4
    split_scans: false
  binazhwp_synchronous:
    bins: { n_bins: 4, interpolate: false, smooth: false }
    n_harmonics: 4
  binaz_synchronous:    # new!
    bins: { n_bins: 4, interpolate: false, smooth: false }
```

## Behavioural changes

- Template amplitude outputs are now structured `(n_dets, *shape)`, not flat `(n_params,)`.
- `azhwp_synchronous` honours `legendre.min_order` (previous operator always used `0..n-1`).
- `template_{key}_*` auxiliary outputs (azimuth-grid fit) are no longer produced.
- Polynomial template uses `SegmentedBasis` instead of dense per-interval blocks (equivalent result).

YAML configs using the old template keys (`max_poly_order`, `n_polynomials`, `n_azimuth_bins`, …) must move to the new `legendre`/`bins` structure (see above)

### CODE REVIEW

File: src/furax/mapmaking/config.py [hunk 1]
```diff
@@ -223,63 +223,105 @@ def __post_init__(self) -> None:
             raise ValueError('exactly one of healpix or wcs must be set.')
 
 
+class LegendreOrders(NamedTuple):
```
Comment (Wuhyun) [inline src/furax/mapmaking/config.py]: Can this be called `PolynomialOrders` instead, since it's not limited to legendre polynomials?

File: src/furax/mapmaking/config.py [hunk 2]
```diff
@@ -223,63 +223,105 @@ def __post_init__(self) -> None:
...
+class PolynomialConfig:
+    legendre: LegendreOrders = LegendreOrders(0, 3)
+    """Legendre orders for the polynomial drift template."""
+
+
+@dataclass
+class ScanSynchronousConfig:
+    """Scan-synchronous (azimuth-only) signal on a global Legendre basis.
+
+    For constant-elevation scans the pickup is a function of azimuth only.
```
Comment (Wuhyun) [inline src/furax/mapmaking/config.py]: "For constant-elevation scans the pickup is a function of azimuth only"
->
"Represents signals that only depends on the telescope's azimuth"
maybe?

File: src/furax/mapmaking/config.py [hunk 3]
```diff
@@ -223,63 +223,105 @@ def __post_init__(self) -> None:
...
+    See `BinAzSynchronousConfig` for the binned (sotodlib azss) variant.
```
Comment (Wuhyun) [inline src/furax/mapmaking/config.py]: I think sotodlib's azss template is different from ours; they still fit a smooth polynomial to the az-binned data, whereas our binazss and binazhwp ones remain binned and discontinuous

File: src/furax/mapmaking/templates.py [hunk 4]
```diff
@@ -17,773 +40,567 @@
...
+        n = len(self.shape)
+        idx = tuple(range(n))
+        return jnp.einsum(self.values, (*idx, n), signal, (n,), idx)
 
-        # Again, we can't run below due to varying block sizes
-        # block_mvs = jnp.einsum('ijk,ikl->ijl', params, self.blocks)
 
-        return jnp.concatenate(block_mvs, axis=1)
+class DecimatedTensorBasis(Basis):
+    """Dense basis stored on a coarser time grid to save memory.
```
Comment (Wuhyun) [inline src/furax/mapmaking/templates.py]: Is TensorBasis equivalent to DecimatedTensorBasis with decimate=1 ? If so, would it be worth absorbing the latter into the former?

File: src/furax/mapmaking/templates.py [hunk 5]
```diff
@@ -17,773 +40,567 @@
...
-        """Compute optional data to be computed at the end for the best-fit template amplitudes."""
-        return {}
+    @classmethod
+    def temperature(
+        cls,
+        temperature: Float[Array, 'det samp'],
+        dtype: DTypeLike,
+        fit_band: tuple[float, float] | None = None,
+        sample_rate: Float[Array, ''] | float = 1.0,
+        decimate: int = 1,
```
Comment (Wuhyun) [inline src/furax/mapmaking/templates.py]: `decimate` -> `decimation_factor` or `decimate_q` might be clearer. Currently sounds like a boolean variable...

File: src/furax/mapmaking/templates.py [hunk 6]
```diff
@@ -17,773 +40,567 @@
...
-        templates: Float[Array, 'mode samp'],
-    ):
-        # TODO: add checks
-        super().__init__(n_params, n_dets, n_samps, dtype)
-        object.__setattr__(self, 'n_modes', n_modes)
-        object.__setattr__(self, 'templates', templates)
+    ) -> Self:
+        """Scan-synchronous (azimuth-only) template on a global Legendre basis."""
+        legs = _legendre(azimuth, legendre.min_order, legendre.max_order, dtype)
+        return cls.from_basis(TensorBasis.create(legs), n_dets=n_dets)
```
Comment (Wuhyun) [inline src/furax/mapmaking/templates.py:424]: Neat!

File: src/furax/mapmaking/templates.py [hunk 7]
```diff
@@ -17,773 +40,567 @@
...
-
-    n_azimuth_bins: int = field(metadata={'static': True})
-    n_harmonics: int = field(metadata={'static': True})
-    bin_templates: Float[Array, 'bin samp']
-    harm_templates: Float[Array, 'harm samp']
+        poly = _legendre(azimuth, legendre.min_order, legendre.max_order, dtype)
+        if scan_mask is not None:
+            poly = scan_mask[None, :] * poly
+        harm = _harmonics(hwp_angles, n_harmonics, dtype, dc=True)
+        return cls.from_basis(KroneckerBasis.create((poly, harm)), n_dets=n_dets)
```
Comment (Wuhyun) [inline src/furax/mapmaking/templates.py:473]: Very neat!

**Review by Wuhyun — DISMISSED** [review]
Neat way of categorising and handling basis functions for templates, thanks a lot! Minor comments below.

File: src/furax/mapmaking/templates.py (hunk 4 above)
Comment (sbiquard) [reply src/furax/mapmaking/templates.py]: It is! Absorbing it makes sense, even simplifies a few things. Consider it done!
